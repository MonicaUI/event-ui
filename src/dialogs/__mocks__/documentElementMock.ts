export const getMockedMessageContainer = (): $TSFixMe => {
  // Mock a DOM to play around
  document.body.innerHTML = `
    <div>
    <div name="messageContainer" tabIndex=-1>message container</div>
    </div>
  `;
  const mockMessageContainerFocus = jest.fn();
  const mockMessageContainer = document.getElementsByName('messageContainer')[0];
  mockMessageContainer.focus = mockMessageContainerFocus;
  return mockMessageContainer;
};

export const getMockedFieldInputs = (fieldNames: $TSFixMe): $TSFixMe => {
  // Mock a DOM to play around
  document.body.innerHTML = `
    <div>
    ${fieldNames.map(fieldName => `<input name=${fieldName}>input-${fieldName}</input>`)}
    </div>
  `;
  const mockInputFocus = jest.fn();
  const mockInputContainers = [];
  fieldNames.forEach(fieldName => {
    const mockInputContainer = document.getElementsByName(fieldName)[0];
    mockInputContainer.focus = mockInputFocus;
    mockInputContainers.push(mockInputContainer);
  });
  return mockInputContainers;
};
