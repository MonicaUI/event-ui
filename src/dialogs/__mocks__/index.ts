/* eslint-env jest */
import noopAction from '../../testUtils/noopAction';

export const openAlreadyRegisteredDialog = jest.fn(noopAction);

export const openDeclineRegistrationDialog = jest.fn(noopAction);

export const openStartNewRegistrationDialogFromPageLanding = jest.fn(noopAction);

export const openStartNewRegistrationDialogDuringRegistration = jest.fn(noopAction);

export const startNewRegistrationAndNavigateToRegistration = jest.fn(noopAction);

export const openEventWaitlistDialog = jest.fn(noopAction);

export const openEventTemporaryClosedErrorDialog = jest.fn(noopAction);

export const openNoAdmissionItemAvailableForRegistrationTypeDialog = jest.fn(noopAction);

export const openEventStatusDialog = jest.fn(noopAction);

export const openPrivateEventErrorDialog = jest.fn(noopAction);

export const openRegistrationInOtherTabDialog = jest.fn(noopAction);

export const openKnownErrorDialog = jest.fn(noopAction);

export const openPaymentProcessingErrorDialog = jest.fn(noopAction);

export const openInformationAlreadyUsedDialog = jest.fn(noopAction);

export const openPartialRegistrationConfirmationDialog = jest.fn(noopAction);

export const openPreviewModeWarningDialog = jest.fn(noopAction);

export const openCapacityReachedDialog = jest.fn(noopAction);

export const openGroupCancelRegistrationDialog = jest.fn(noopAction);

export const openSessionOverlapWarningDialog = jest.fn(noopAction);

export const openGroupMemberRemoveDialog = jest.fn(noopAction);

export const openSharePromptDialog = jest.fn(noopAction);

export const openConfirmationDialog = jest.fn(noopAction);

export const openGuestProductSelectionDialog = jest.fn(noopAction);

export const openGroupRegistrationTypeDialog = jest.fn(noopAction);

export const openGuestDetailsDialog = jest.fn(noopAction);

export const openGuestRemoveDialog = jest.fn(noopAction);

export const openContactPlannerDialog = jest.fn(noopAction);

export const openCancelRegistrationSuccessConfirmationDialog = jest.fn(noopAction);

export const openCancelRegistrationDialog = jest.fn(noopAction);

export const openShoulderDateApprovalWarningDialog = jest.fn(noopAction);

export const openTransactionInProcessingErrorDialog = jest.fn(noopAction);
