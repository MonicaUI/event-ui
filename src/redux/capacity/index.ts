import baseReducer, {
  loadCapacityCounts,
  getAdmissionItemCapacityIds,
  getRegTypeCapacityIds,
  getAllSessionCapacityIds,
  getQuantityItemCapacityIds,
  loadQuantityItemCapacity,
  loadSessionSelectionCapacityCounts,
  loadEventRegistrationCapacityCounts,
  checkIfLimitedCountForSingleCapacity
} from 'event-widgets/redux/modules/capacity';
import { getRegPackId } from '../selectors/shared';
import { UPDATE_REG_CART_SESSION_SUCCESS } from '../registrationForm/regCart/actionTypes';
import { refPreserving } from '@cvent/ref-preserving-function';

const collectAllCapacityIdsFromStore = (event, regPackId) => {
  const regTypeCapacityIds = getRegTypeCapacityIds(event, null, regPackId);
  const sessionCapacityIds = getAllSessionCapacityIds(event);
  const admissionCapacityIds = getAdmissionItemCapacityIds(event);
  const quantityItemCapacityIds = getQuantityItemCapacityIds(event);
  const eventCapacityId = event.capacityId;
  return [
    ...sessionCapacityIds,
    ...admissionCapacityIds,
    eventCapacityId,
    ...regTypeCapacityIds,
    ...quantityItemCapacityIds
  ];
};

export const fetchCapacitySummaries = async (
  regCartClient: $TSFixMe,
  capacityClient: $TSFixMe,
  accessToken: $TSFixMe,
  regCartId: $TSFixMe,
  capacityIds: $TSFixMe,
  hasSessionInitialized: $TSFixMe
): Promise<$TSFixMe> => {
  // If regCartId is present, call the regAPI capacity endpoint, otherwise just call capacity service directly
  let capacity = null;
  if (regCartId && hasSessionInitialized) {
    capacity = regCartClient.getCapacitySummaries(accessToken, regCartId, capacityIds);
  } else if (regCartId && !hasSessionInitialized) {
    capacity = regCartClient.getCapacitySummariesOnPageLoad(accessToken, regCartId, capacityIds);
  } else {
    capacity = capacityClient.getCapacitySummaries(accessToken, capacityIds);
  }
  return capacity;
};

export const loadAvailableCapacityCounts = () => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      event,
      clients: { regCartClient, capacityClient },
      userSession
    } = getState();
    const capacityIds = collectAllCapacityIdsFromStore(event, getRegPackId(getState()));
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const regCartId = userSession && userSession.regCartId;
    dispatch(
      loadCapacityCounts(
        await fetchCapacitySummaries(regCartClient, capacityClient, accessToken, regCartId, capacityIds, true)
      )
    );
  };
};

export const loadAvailableEventRegistrationCapacityCounts = () => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      event,
      clients: { regCartClient, capacityClient },
      userSession
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const regCartId = userSession && userSession.regCartId;
    const regTypeCapacityIds = getRegTypeCapacityIds(event, null, getRegPackId(getState()));
    const eventCapacityId = event.capacityId;
    const capacityIds = [...regTypeCapacityIds, eventCapacityId];
    dispatch(
      loadEventRegistrationCapacityCounts(
        await fetchCapacitySummaries(regCartClient, capacityClient, accessToken, regCartId, capacityIds, true)
      )
    );
  };
};

export const loadAvailableQuantityItemCapacityCounts = () => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      event,
      clients: { regCartClient, capacityClient },
      userSession
    } = getState();
    const quantityItemCapacityIds = getQuantityItemCapacityIds(event);
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const regCartId = userSession && userSession.regCartId;
    dispatch(
      loadQuantityItemCapacity(
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 0 arguments, but got 1.
        await fetchCapacitySummaries(
          regCartClient,
          capacityClient,
          accessToken,
          regCartId,
          quantityItemCapacityIds,
          true
        )
      )
    );
  };
};

/**
 * exclude those session capacity id which has unlimited capacity
 * this is to reduce the number of request in capacity service for session capacity
 * e.g: when <loadAvailableSessionCapacityCounts> is being called
 */
const excludeUnlimitedSessionCapacityIds = (sessionCapacityIds, eventCapacity) => {
  if (eventCapacity) {
    return (
      // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
      sessionCapacityIds &&
      sessionCapacityIds.filter(sessionCapacityId =>
        checkIfLimitedCountForSingleCapacity(sessionCapacityId, eventCapacity)
      )
    );
  }
  return sessionCapacityIds;
};

export const loadAvailableSessionCapacityCounts = (capacityIds = null) => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      clients: { regCartClient, capacityClient },
      userSession,
      event,
      capacity
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const regCartId = userSession && userSession.regCartId;
    let sessionCapacityIds = capacityIds;
    if (!sessionCapacityIds) {
      sessionCapacityIds = getAllSessionCapacityIds(event);
    }
    sessionCapacityIds = excludeUnlimitedSessionCapacityIds(sessionCapacityIds, capacity);
    /*
     * if sessionCapacityIds collection doesn't have anything, we don't need to make the capacity service call
     * e.g: all sessions are unlimited
     *
     * In case of no sessions present
     * capacity service call breaks when agenda widget is dropped on page
     */
    if (sessionCapacityIds && sessionCapacityIds.length > 0) {
      dispatch(
        loadSessionSelectionCapacityCounts(
          await fetchCapacitySummaries(regCartClient, capacityClient, accessToken, regCartId, sessionCapacityIds, true)
        )
      );
    }
  };
};

/**
 * initially when user lands on guest side summary page, we are not including capacities for sessions
 * from service side rendering, when user navigates to agenda widget tab, we need to load capacities for
 * all available sessions from event, since currently the capacity in state would not include for sessions
 */
export const loadAvailableSessionCapacityCountsForAgendaWidget = () => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      clients: { regCartClient, capacityClient },
      userSession,
      event
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const regCartId = userSession && userSession.regCartId;
    const sessionCapacityIds = getAllSessionCapacityIds(event);

    if (sessionCapacityIds && sessionCapacityIds.length > 0) {
      dispatch(
        loadSessionSelectionCapacityCounts(
          await fetchCapacitySummaries(regCartClient, capacityClient, accessToken, regCartId, sessionCapacityIds, true)
        )
      );
    }
  };
};

function reducer(oldState, action) {
  const state = baseReducer(oldState, action);
  switch (action.type) {
    case UPDATE_REG_CART_SESSION_SUCCESS: {
      const updates = {};
      for (const { capacityId, change } of action.payload.optimisticCapacityUpdates) {
        if (state[capacityId] && state[capacityId].totalCapacityAvailable !== -1) {
          updates[capacityId] = {
            ...state[capacityId],
            availableCapacity: state[capacityId].availableCapacity + change
          };
        }
      }
      return {
        ...state,
        ...updates
      };
    }
    default:
      return state;
  }
}

export default refPreserving(reducer);
