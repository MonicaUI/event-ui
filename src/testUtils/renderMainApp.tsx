import React from 'react';
import { render } from '@testing-library/react';
import startAppRender from '../main';
import EventSnapshotClient from '../clients/EventSnapshotClient';
import RegCartClient from '../clients/RegCartClient';
import LookupClient from 'event-widgets/clients/LookupClient';
import CapacityClient from 'event-widgets/clients/CapacityClient';
import AppointmentsClient from 'event-widgets/clients/AppointmentsClient';
import { startIntercepting, finishIntercepting } from 'react-dom';
import myHistory from '../myHistory';
import { act } from '@testing-library/react';
import EventGuestClient from '../clients/EventGuestClient';

jest.mock('event-widgets/redux/modules/text', () => {
  const translate = (resx, options = null) => (options ? `${resx}:${JSON.stringify(options)}` : resx);
  const initialState = {
    translate,
    translateWithDatatags: translate,
    translateDate(date) {
      return date.toISOString();
    },
    translateTime(date) {
      return date.toISOString();
    },
    locale: 'mock-locale',
    resolver: {
      registerTranslations: () => () => undefined
    }
  };
  const reducer = () => initialState;
  return reducer;
});
jest.mock('react-dom', () => {
  const ReactDOM = jest.requireActual<$TSFixMe>('react-dom');
  const realRender = ReactDOM.render;
  let interceptRenders = false;
  let interceptedComponents = [];
  ReactDOM.startIntercepting = () => {
    interceptRenders = true;
    interceptedComponents = [];
  };
  ReactDOM.finishIntercepting = () => {
    const result = interceptedComponents;
    interceptRenders = false;
    interceptedComponents = [];
    return result;
  };
  ReactDOM.render = (...args) => {
    if (interceptRenders) {
      interceptedComponents.push(args[0]);
      return;
    }
    return realRender(...args);
  };
  return ReactDOM;
});

jest.mock('../clients/EventSnapshotClient', () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const EventSnapshot = require('../../fixtures/EventSnapshot.json');
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const EventTravelAirportsData = require('../../fixtures/travelAirportsData.json');
  EventSnapshot.eventSnapshot.version = 'eventSnapshotVersion';
  class MockEventSnapshotClient {
    static _eventModifier = x => x;
    async getEventSnapshot() {
      return MockEventSnapshotClient._eventModifier(EventSnapshot.eventSnapshot);
    }
    async getAccountSnapshot() {
      return EventSnapshot.accountSnapshot;
    }
    async getEventTravelAirports() {
      return EventTravelAirportsData;
    }
    async getEventTravelSnapshot() {
      return {};
    }
    static setEventModifier(f) {
      MockEventSnapshotClient._eventModifier = f;
    }
    async getVisibleProducts() {
      return {};
    }

    async getRegCartVisibleProducts() {
      return {};
    }
  }
  return MockEventSnapshotClient;
});
jest.mock('../clients/EventGuestClient');
jest.mock('../clients/EventEmailClient', () => {
  class MockEventEmailClient {
    async getSubscriptionStatus() {
      return {};
    }
    async setSubscriptionStatus() {
      return {};
    }
  }
  return MockEventEmailClient;
});
jest.mock('event-widgets/clients/LookupClient');
jest.mock('event-widgets/clients/CapacityClient');
jest.mock('event-widgets/clients/AppointmentsClient');
jest.mock('../clients/RegCartClient');
jest.mock('nucleus-core/dialog/Dialog', () => {
  // eslint-disable-next-line react/prop-types
  return ({ isOpen, header, children, actions }) => {
    return isOpen ? (
      <div className="mockNucleusCoreDialog">
        <h1>{header}</h1>
        <div>{children}</div>
        {actions && <div>{actions}</div>}
      </div>
    ) : null;
  };
});
jest.mock('nucleus-guestside-site/src/containers/Page', () => {
  // eslint-disable-next-line react/prop-types
  return ({ page }) => <div className="mockPage">{page.id}</div>;
});
jest.mock('nucleus-widgets/renderers/readOnlyContent', () => {
  // eslint-disable-next-line react/prop-types
  return ({ rootLayoutItemId }) => <div className="mockReadOnlyContent">{rootLayoutItemId}</div>;
});
jest.mock('@cvent/nucleus-remote-log-client', () => {
  class MockRemoteLogClient {
    debug() {}
    info() {}
    warn() {}
    error() {}
  }
  return MockRemoteLogClient;
});
// @ts-expect-error ts-migrate(2540) FIXME: Cannot assign to 'localStorage' because it is a re... Remove this comment to see the full error message
window.localStorage = {
  getItem() {},
  setItem() {}
};
window.scrollTo = () => {};
jest.mock('../myHistory', () => {
  const createHistory = jest.requireActual<$TSFixMe>('history').createMemoryHistory;
  return createHistory();
});
jest.mock('nucleus-widgets/utils/browserFeatureDetection', () => {
  return () => Promise.resolve(false);
});
function mockCodeSnippets() {
  window.CVENT = {};
  window.CVENT.runTriggerHandlers = jest.fn();
}

const fetchMockAssets = async clients => {
  return Promise.all([
    clients.eventSnapshotClient.getAccountSnapshot(),
    clients.eventSnapshotClient.getEventSnapshot(),
    clients.eventSnapshotClient.getEventTravelSnapshot(),
    clients.lookupClient.getTimezone(null, []),
    clients.lookupClient.getCurrencies(),
    clients.capacityClient.getCapacitySummaries(),
    clients.eventSnapshotClient.getEventTravelAirports(),
    clients.appointmentsClient.getApptEventDetails()
  ]);
};

const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

export default async function renderMainApp(
  appSettings: $TSFixMe,
  startingPath: $TSFixMe,
  eventModifier: $TSFixMe
): Promise<$TSFixMe> {
  let result;
  await act(async () => {
    mockCodeSnippets();
    (EventSnapshotClient as $TSFixMe).setEventModifier(eventModifier);
    const mockClients = {
      eventSnapshotClient: new EventSnapshotClient(),
      lookupClient: new LookupClient(),
      regCartClient: new RegCartClient(),
      capacityClient: new CapacityClient(),
      appointmentsClient: new AppointmentsClient(),
      eventGuestClient: new EventGuestClient()
    };
    const [
      accountSnapshot,
      eventSnapshot,
      travelSnapshot,
      eventTimezoneResponse,
      currencyResponse,
      capacity,
      airportsResponse
    ] = await fetchMockAssets(mockClients);

    await act(async () => {
      myHistory.push(startingPath);
      await wait(0);
    });
    startIntercepting();

    await startAppRender(
      {
        ...appSettings,
        cultureCode: 'en',
        eventSnapshotVersion: 'eventSnapshotVersion',
        capacityIds: []
      },
      {
        clients: mockClients,
        assets: {
          accountSnapshot,
          eventSnapshot,
          travelSnapshot,
          eventTimezone: eventTimezoneResponse,
          currencies: currencyResponse.currencies,
          capacity,
          airports: airportsResponse.airports
        }
      },
      // @ts-expect-error ts-migrate(2554) FIXME: Expected 2 arguments, but got 3.
      name
    );
    const [, mainAppComponent] = finishIntercepting();
    if (!mainAppComponent) {
      throw new Error('main did not render anything');
    }
    result = render(mainAppComponent);
  });
  result.getDOMNode = () => result.container.querySelector('div');
  return result;
}
