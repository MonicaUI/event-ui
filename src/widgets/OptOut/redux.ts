/* eslint-disable import/no-cycle */

import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Opt-out and opt-in Redux module:
 * Event invitees can click "Opt Out" link in their emails to opt them out all communications from the event account
 */
const GET_OPT_OUT_STATUS = 'event-guestside-site/optOut/GET_OPT_OUT_STATUS';
const OPT_OUT_CONTACT = 'event-guestside-site/optOut/OPT_OUT_CONTACT';

const createGetOptedOutStatusAction = contactInfo => {
  return {
    type: GET_OPT_OUT_STATUS,
    payload: contactInfo
  };
};

export const getOptOutStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    const optedOutContact = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, guestId, contactId);
    dispatch(createGetOptedOutStatusAction(optedOutContact));
  };

export const createOptedOutAction = (contactInfo: $TSFixMe): $TSFixMe => {
  return {
    type: OPT_OUT_CONTACT,
    payload: contactInfo
  };
};

export const optOutAction =
  (optOutStatus: $TSFixMe, fromConfirmPage: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    // eslint-disable-next-line import/no-cycle
    const page = (await import(/* webpackChunkName optOut */ './content')).optInOutPageNew;
    if ((fromConfirmPage && optOutStatus) || !fromConfirmPage) {
      const optedOutContact = await eventGuestClient.optOut(
        accessToken,
        eventId,
        inviteeId,
        guestId,
        contactId,
        optOutStatus
      );
      dispatch(createOptedOutAction(optedOutContact));
    }
    dispatch(injectPageSnapshot(page));
  };

export const prepareForOptOutPageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    let page;
    if (!inviteeId && !guestId && !contactId) {
      page = (await import(/* webpackChunkName optOut */ './invalidInviteeContent')).invalidInviteeOptOut;
    } else {
      await dispatch(getOptOutStatus());
      if (getState() && getState().optOut && getState().optOut.optOutStatus) {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).optInOutPage;
      } else {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).confirmationPage;
      }
    }
    dispatch(injectPageSnapshot(page));
  };

export const optOutReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_OPT_OUT_STATUS:
    case OPT_OUT_CONTACT:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
