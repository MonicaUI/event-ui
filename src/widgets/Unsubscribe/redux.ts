import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Unsubscribe and re-subscribe Redux module:
 * Event invitees can click "Unsubscribe" link in their emails to opt them out from the event emails communication
 */
const GET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/GET_UNSUBSCRIBE_STATUS';
const SET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/SET_UNSUBSCRIBE_STATUS';

const createGetUnsubscribeStatusAction = (unsubscribedContact, optOutStatus) => {
  return {
    type: GET_UNSUBSCRIBE_STATUS,
    payload: {
      ...optOutStatus,
      unsubscribeStatus: !unsubscribedContact.subscribed
    }
  };
};

export const getUnsubscribeStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient, eventGuestClient },
      accessToken,
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const unsubscribedContact = await eventEmailClient.getSubscriptionStatus(eventId, inviteeId, encodedGuestId);
    // optOutStatus object has invitee/guest and planner infomation for display
    const optOutStatus = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, encodedGuestId);
    dispatch(createGetUnsubscribeStatusAction(unsubscribedContact, optOutStatus));
  };

const createUnsubscribeAction = unsubscribeStatus => {
  return {
    type: SET_UNSUBSCRIBE_STATUS,
    payload: { unsubscribeStatus }
  };
};

export const unsubscribeAction =
  (unsubscribeStatus: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient },
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const subscribed = !unsubscribeStatus;
    await eventEmailClient.setSubscriptionStatus(eventId, inviteeId, encodedGuestId, subscribed);
    dispatch(createUnsubscribeAction(unsubscribeStatus));
  };

export const prepareForUnsubscribePageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    let unsubscribePage;
    if (!inviteeId && !encodedGuestId) {
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './invalidInviteeContent'))
        .invalidInviteeUnsubscribePage;
    } else {
      await dispatch(getUnsubscribeStatus());
      // eslint-disable-next-line import/no-cycle
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './content')).unsubscribePage;
    }
    dispatch(injectPageSnapshot(unsubscribePage));
  };

export const unsubscribeReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_UNSUBSCRIBE_STATUS:
    case SET_UNSUBSCRIBE_STATUS:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
