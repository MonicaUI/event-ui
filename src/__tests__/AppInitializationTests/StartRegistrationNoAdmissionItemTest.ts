import renderMainApp from '../../testUtils/renderMainApp';
import * as eventStatus from 'event-widgets/clients/EventStatus';
import RegCartClient from '../../clients/RegCartClient';

// eslint-disable-next-line jest/no-mocks-import
import { getMockedMessageContainer } from '../../dialogs/__mocks__/documentElementMock';
getMockedMessageContainer();

const mockEventId = '11111111-2222-3333-4444-555555555555';
const eventId = mockEventId;
const fakePath = `/${eventId}/`;

jest.mock('../../clients/RegCartClient', () => {
  const { setIn } = jest.requireActual<$TSFixMe>('icepick');
  function MockRegCartClient() {}
  MockRegCartClient.prototype.createRegCart = jest.fn((authToken, regCart) => {
    return {
      regCart: {
        ...regCart,
        regCartId: 'regCartId',
        eventRegistrations: setIn(
          regCart.eventRegistrations,
          ['00000000-0000-0000-0000-000000000001', 'registrationPathId'],
          'todoThisShouldBeRegPathId'
        ),
        eventSnapshotVersions: {
          [mockEventId]: 'eventSnapshotVersion'
        }
      },
      validationMessages: []
    };
  });
  return MockRegCartClient;
});

jest.mock('../../redux/capacity');
// eslint-disable-next-line @typescript-eslint/no-var-requires
require('../../redux/capacity').loadAvailableCapacityCounts.mockImplementation(() => () => {});

const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

test('Shows the no admission item dialog, when creating a reg cart, if that validation is present', async () => {
  (RegCartClient.prototype.createRegCart as $TSFixMe).mockImplementation(() => {
    // eslint-disable-next-line no-throw-literal
    throw {
      responseStatus: 422,
      responseBody: {
        validationMessages: [
          {
            severity: 'Error',
            localizationKey: 'REGAPI.ADMISSION_ITEM_EXACTLY_ONE'
          }
        ]
      }
    };
  });

  const mainApp = await renderMainApp(
    {
      eventContext: {
        eventId
      },
      plannerRegSettings: {
        exitUrl: '/fakeUrl'
      },
      eventLaunchWizardSettings: '{}'
    },
    fakePath + 'regProcessStep1',
    event => ({
      ...event,
      status: eventStatus.ACTIVE
    })
  );
  await wait(0);
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  expect(require('../../clients/RegCartClient').prototype.createRegCart).toHaveBeenCalled();
  expect(mainApp.getDOMNode()).toMatchSnapshot();
});
