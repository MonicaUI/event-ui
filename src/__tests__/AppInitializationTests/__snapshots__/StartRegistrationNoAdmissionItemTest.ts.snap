// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows the no admission item dialog, when creating a reg cart, if that validation is present 1`] = `
<div>
  <div
    class="wrapper"
  />
  <div
    class="mockNucleusCoreDialog"
  >
    <h1 />
    <div>
      <div
        class="dialogContainer"
      >
        <div
          class="dialogChildren"
        >
          <div>
            <div
              aria-describedby="dialogHeader"
              class=""
              style="background-color: rgb(221, 226, 230); border-radius: 0;"
            >
              <div
                class=""
                id="dialogHeader"
                style="background-color: rgb(249, 32, 242); padding: 15px 1.5625rem 15px 1.5625rem;"
              >
                <h2
                  class=""
                  style="color: rgb(255, 255, 255); font-family: Comic Sans MS, Comic Sans, Chalkboard, Marker Felt, sans-serif; font-size: 3.5625rem; font-style: normal; font-weight: 700; line-height: 1.3;"
                >
                  <div
                    class=""
                    data-cvent-id="close"
                    role="button"
                    tabindex="0"
                  >
                    <span
                      class="container"
                    >
                      <span
                        aria-label="EventGuestSide_StandardDialog_FallbackText_Close__resx"
                        class=""
                        role="img"
                      />
                    </span>
                  </div>
                  EventGuestSide_RegistrationTypeConflict_Title_resx
                </h2>
              </div>
            </div>
            <div
              class=""
              name="messageContainer"
              style="background-color: rgb(243, 59, 21); padding: 0px 1.25rem 1.875rem 1.25rem;"
              tabindex="-1"
            >
              <span
                class="container"
              >
                <span
                  aria-hidden="true"
                  class=""
                  role="img"
                />
              </span>
              <div
                class=""
                style="font-family: PT Sans, sans-serif; font-weight: 400; font-size: 2.5rem; color: rgb(39, 63, 105); line-height: 1.3; font-style: normal; background-color: transparent; border-radius: 0; padding: 0px 0px 0px 0px;"
              >
                EventGuestSide_RegistrationTypeConflict_Title_resx
              </div>
              <p
                class=""
              >
                EventGuestSide_RegistrationTypeError_NoAdmissionItemAvailableHelpText_resx
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</div>
`;
